### ISSUE TITLE
[google_maps_flutter] API for setting GMSServices.provideAPIKey

### ISSUE THREAD

**Issue by awhitford** [flutter/flutter#51433]
Can we get an API for setting `GMSServices.provideAPIKey` so that we can avoid baking hard-coded values in `AppDelegate` (as described [here](https://pub.dev/packages/google_maps_flutter#ios))?

## Use case

Avoid hard-coding sensitive API keys in `AppDelegate` (as described [here](https://pub.dev/packages/google_maps_flutter#ios)).

## Proposal

If there was an API exposed to Dart, one could get the API key from the database after authentication or _Firebase Remote Config_, and then pass it to _Google Maps_ before instantiating a widget.

_Will this approach work for [Android](https://pub.dev/packages/google_maps_flutter#android) too?_  (Obviously, a uniform solution is ideal.)

**Comment by cielo** [comment]
It is possible to restrict usage of API keys to specific Android app, iOS App, HTTP Referer, or IP addresses.

https://console.cloud.google.com/apis/credentials

I believe authenticating with Firebase Remote Config itself will require API key, because your app will need to initialize Firebase clients with certain key. (usually available in GoogleService-Info.plist)

So I am unsure if putting Google Maps API key (which will likely be same as the Firebase API key if you are using One API key for the iOS app) in firebase will make this safer, unless you restrict app key itself.

By the way, it is recommended not to store any sensitive credential value in Firebase remote config.

**Comment by awhitford** [comment]
> By the way, it is recommended not to store any sensitive credential value in Firebase remote config.

It is also [not advisable to store API Keys in source code](https://developers.google.com/maps/api-key-best-practices#no_store_apikey).  So, if you can't store them in source code, and you can't use a service like _Firebase Remote Config_, what is one supposed to do?

**Comment by cielo** [comment]
> > By the way, it is recommended not to store any sensitive credential value in Firebase remote config.
> 
> It is also [not advisable to store API Keys in source code](https://developers.google.com/maps/api-key-best-practices#no_store_apikey). So, if you can't store them in source code, and you can't use a service like _Firebase Remote Config_, what is one supposed to do?

As I have said before, 

>It is possible to restrict usage of API keys to specific Android app, iOS App, HTTP Referer, or IP addresses.
>
>https://console.cloud.google.com/apis/credentials

So your Google Cloud API key can be put into configuration file in app package (e.g. .plist file in iOS). Even if someone steals it, they won't be able to use it because API key will only work under given app, and Google Cloud services will reject requests not coming from your app. 

The link that you are referring to is talking about private key or API key which cannot be restricted by client type, id.

There is no point in persisting sensitive information in RemoteConfig, as anyone can see the full config value.

**Comment by awhitford** [comment]
Thank you @cielo, I am reverting to hard-coding the API Key in `AppDelegate.swift`, following the [instructions](https://pub.dev/packages/google_maps_flutter#ios).

My reservations were driven by the documentation I cited, as well comments from #51267 that literally said, _"API key in your source code is only meant for testing purpose in documentation not for production release."_

**Comment by cielo** [comment]
That should be fine. The [same doc](https://developers.google.com/maps/api-key-best-practices#no_store_apikey) that you linked has this information.

>**Restrict your API keys**. You can best protect your API key by restricting it to specific IP addresses, referrer URLs or mobile apps, and specific APIs, as this significantly reduces the impact of a key compromise.
>You can specify application and API restrictions for a key from the console by opening the Credentials page and then either creating a new API key with the settings you want, or editing the settings of an API key. See section Restricting API keys for full details.

So if you are hard coding google maps key, make sure it is only useable in your iOS app bundle id. By default, they don't impose limits and anyone who obtains your key can abuse it.

**Comment by AlexHartford** [comment]
@cielo What about service accounts? I need to integrate Dialogflow with my Flutter app, and would really prefer not to have a streaming backend server just to protect my credentials... especially when the rest of my app is using Firebase.  As it stands, I have to include a JSON file for my service account in my app bundle.  If I could restrict the usage of the service account similar to how you can restrict API keys, it would help alleviate this concern.  Currently, there is no suitable way to securely provide this data in Flutter for a production application.  Many people are asking these types of questions without a suitable answer.

**Comment by cielo** [comment]
@AlexHartford If you are talking about private credential JSON file that is generated for service account, you should never put that in your app package. Those private keys are meant for secure environments, such as your own server environments, and that is just how those keys work. You shouldn't embed service account key in website or in app. 

> Currently, there is no suitable way to securely provide this data in Flutter for a production application.

There is no way to securely store credential data in ANY of the Flutter/Android/iOS/Web/Xarmin application. Any savvy individual will be able to extract credential file from the app package if they want to. I am unsure why anyone would think there is such a way. And above is the reason why anyone won't be able to find an answer like you said.

> Many people are asking these types of questions without a suitable answer.

Even the Dialogflow [asks](https://dialogflow.com/docs/reference/v2-auth-setup) you to store private key in a safe place.

> _You can only download this JSON file once, so make sure to save the file and keep it somewhere safe. If you lose this key or it becomes compromised, you can use the same process to create another._

So, you would never put those private key in an app package which will be distributed to tens of thousands of users.

So what would you have to do? You must setup server environment which keeps track of service account keys (if you use GCE, this step may be a lot simpler) and authenticate requests. You cannot just say 'I prefer not to have a streaming backend server just to protect my credentials'., as that is a necessary step to protect your information.

**Comment by AlexHartford** [comment]
@cielo I appreciate the quick reply.  I do believe there should be a more elegant solution to this problem though.  Why even bother providing https://pub.dev/packages/googleapis (with no mention that including service accounts in-code is not suitable for production)? 

Yes, most developers understand this is wrong - but the package linked above, provided by dart.dev, breaks this rule fundamentally.  And I can tell from the usage of this package and its derivatives (e.g. https://pub.dev/packages/flutter_dialogflow) many people are just letting this security gap slide.  There are 2 medium articles ([1](https://medium.com/flutter-community/build-a-chatbot-in-20-minutes-using-flutter-and-dialogflow-8e9af1014463), [2](https://medium.com/flutterdevs/chatbot-in-flutter-using-dialogflow-70e28665a827)) that have gained exponentially more traction than the [project I linked in my previous comment](https://medium.com/google-cloud/building-your-own-conversational-voice-ai-with-dialogflow-speech-to-text-in-web-apps-part-i-b92770bd8b47) that are making use of this flutter_dialogflow package.  So clearly, people want this to be easier. 

Having to set up a server for no value-added other than hiding my credentials? Trust me, I understand that credentials being private is fundamentally important. Yes, it is a necessary step to protect my information, **currently**.  

Why can't service accounts be restricted in the same way that API keys are?  Can I use an API key instead of a service account for accessing Dialogflow?  I don't have answers to these questions because documentation _does not exist_ on these topics. 

TL;DR: Why have https://pub.dev/packages/googleapis at all if you can't utilize it without breaking security constraints?

**Comment by cielo** [comment]
I will avoid lengthy discussion after this last comment as we are talking about topic that is barely related to what the original requestor of this issue initially asked about.

@AlexHartford 
Service account is not the only way to use https://pub.dev/packages/googleapis. It works well with API keys & credentials method for many of services that it supports. 

From [GitHub repo](https://github.com/dart-lang/googleapis)
> In addition, based on the API being used, you need either an API key or a Client ID. And finally, depending on what kind of application is being used, you **might** need a Service Account.

Do keep in mind that `pub.dev` is a repo for Dart, not just for Flutter.

So it is not right to say 

> Why even bother providing https://pub.dev/packages/googleapis 

Again, it is incorrect to say

>including service accounts in-code is not suitable for production.

because service accounts can be included in-code in a server environment in production.

Also for the examples which need service account private key (e.g. google storage bucket, DialogFlow, etc), the package doesn't mention anything about embedding the key in mobile app. 

The [example](https://pub.dev/packages/googleapis#-example-tab-) shows you how to use the package in dart lang, which does not necessarily mean it is used in flutter. Dart files can run standalone in server environments, and if you have dart server, it should be fine to use that package for DialogFlow. If you want to use `googleapis` dart package in your app, you probably only want to use features that require API key credentials which can be access controlled by bundle id, or app id.

Likewise, there is probably an example out there in Java library that uses DialogFlow private key directly, but that does not mean it is expected to have that same setup in Android. (because Java can run as standalone server, and the library/example was probably meant for java server).

Just because there is a [package](https://pub.dev/packages/flutter_dialogflow#-changelog-tab-) for something does not mean that it is okay to use it. There are probably hundreds of packages listed in other package repo such as Maven(Java) or NPM(JavaScript) that do not follow valid security protocols. That does not mean that users are okay to use them, which is what you are saying here. 

Same concerns can be applied to medium articles ([1](https://medium.com/flutter-community/build-a-chatbot-in-20-minutes-using-flutter-and-dialogflow-8e9af1014463), [2](https://medium.com/flutterdevs/chatbot-in-flutter-using-dialogflow-70e28665a827)) that you are talking about. They are probably written by someone who do not have deep understanding of what it means to embed private key to your app package, or they wrote with an intention that the app will be used by trusted in-network users, or etc. They both received few comments and some reactions, and that is just it. Not sure how that can be interpreted as 'exponentially more people want this feature'. I do see people interested in DialogFlow, which they can do whenever they want since there is a way to do it with a server.

On contrary, the [medium article](https://medium.com/google-cloud/building-your-own-conversational-voice-ai-with-dialogflow-speech-to-text-in-web-apps-part-i-b92770bd8b47) that you mention, which also happens to be written by Google Engineer, correctly shows how architecture should look like, and it uses NodeJS server to relay and authenticate between web app and Dialog APIs.

It is very misleading to say 

> Having to set up a server for no value-added other than hiding my credentials? 

Because with server, you get security, control and gain more access to wide variety of APIs (including DialogFlow).

None of official DialogFlow setup instruction or GCP instruction ask you to embed private key in your mobile app (other than those 2 misleading blog articles), but you are insisting on it because you don't want to setup a server. That is only issue here and nothing else.

e.g. Firebase Admin API which has more powerful features than Firebase Client SDK requires service account key because Admin API is expected to be run in a server environment. Of course, you CAN embed private key in your app and let your app use Firebase Admin API key directly (similar to what you are trying to do with DialogFlow), but that is just poorly designed architecture that poses security risk.

I used Firebase Admin API as example, but there are tens of thousands of APIs (AWS, GCP, Azure, etc) out there that requires API calls to be made in safe server environments, because they rely on private key.

Lastly, do note that `flutter_dialogflow` was not developed or released by google, and it is a third party package developed by another developer. It might be fine to use that package, if your intention is distributing app to trusted small set of users, such as your teammates/coworkers (which I still see as security risk), but if you are thinking about releasing it on App Store or play store, you should never use it unless you want to freely distribute your private keys in the wild. 

Not to mention, when you have to rotate your private keys because of security compromise or for quarterly/annual security review, all users who have old version of app & old private keys will just simply won't be able to use the service.

I see a lot of misinterpretation from your comments as I have pointed them out above. I understand that not everyone is savvy enough, but if you are a developer whose intention is using DialogFlow, one of the core services provided by GCP, and not simple HTML/JS, then you shouldn't be fixated with one path with dead-end. e.g. You will need to understand basics of how private/public crypto key works because service account key that you are keep mentioning is a private key, and you are trying to distribute that key to untrusted users (in a mobile app). You likely need to learn about server programming with whatever language you are familiar with to make this work.

**Comment by AlexHartford** [comment]
I apologize for going quite off-topic but this is ultimately the issue that all the related-but-closed issues led, and you seemed like my best shot at getting some answers.  Thank you for being candid with me.

@cielo I appreciate your detailed response.  You have made a few things clear to me that I was admittedly overlooking because I have only used Dart for Flutter development.  

I have a good amount of experience working with server programming using various languages but was hoping I would be able to avoid it in this case as none of my other functionality requires a server (and I am in a prototyping stage, but still want to follow best-practice where it is logical).

The main point around the articles I brought up was just that Flutter, being relatively young, has some questionable implementations for various things out there.  As I gain more experience with these tools I hope to make positive contributions to the community.

Thanks for addressing me in a straightforward manner.  Keep up the good work.

**Comment by lukehutch** [comment]
> So your Google Cloud API key can be put into configuration file in app package (e.g. .plist file in iOS).

Not so for the `GoogleMap` widget -- the values have to be stored in `AppManifest.xml` on Android, and `AppDelegate.swift` on iOS. Storing an API key in a properties file only works for other Maps APIs, such as Places / Timezone.

There are some fundamental problems with the current approach:
1. Currently there is no way for Flutter code to get access to the Google Maps API key, if you want to reuse that key for calling other Maps APIs. That means you need to define your API key in at least three places, to use both `GoogleMap` and other APIs: `AppManifest`, `AppDelegate`, and your Dart code.
2. The `GoogleMap` widget cannot be configured programmatically from Dart, sourcing the API key from Dart code, the way other Maps APIs can be. The `GoogleMap` widget can only be configured in the respective native environments.

We need either the ability to read the Maps API key from the `AppManifest`/`AppDelegate`, or to configure the `GoogleMap` widget from an API key provided by the Dart API. 

I personally prefer the latter (which would require the key to be configured in only one place, in Dart code), but the former would be an improvement (it would require the key to be configured in two places, for an Android+iOS Flutter app). (Yes, even with the pure Dart solution, there should probably be separate keys for Android and iOS, but they can be switched using `if (Platform.isAndroid)` etc.)
